src/SoftGlobalScope.jl

```diff
@@ -95,16 +116,21 @@ else
     NOTE: `_softscope`` may mutate the `globals` argument (if there are `local` declarations.)
     """
     function _softscope(ex::Expr, globals, locals, insertglobal::Bool=false)
-        if isexpr(ex, :for) || isexpr(ex, :while)
-            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], copy(globals), copy(locals), true))
+        if isexpr(ex, :for)
+            return Expr(ex.head, localassignment(ex.args[1], copy(globals), copy(locals), insertglobal),
+                _softscope(ex.args[2], copy(globals), copy(locals), true))
+        elseif isexpr(ex, :while)
+            return Expr(ex.head, _softscope(ex.args[1], copy(globals), copy(locals), insertglobal),
+                _softscope(ex.args[2], copy(globals), copy(locals), true))
         elseif isexpr(ex, :try)
             try_clause = _softscope(ex.args[1], copy(globals), copy(locals), true)
             catch_clause = _softscope(ex.args[3], copy(globals), ex.args[2] isa Symbol ? union!(locals, ex.args[2:2]) : copy(locals), true)
             finally_clause = _softscope(ex.args[4], copy(globals), copy(locals), true)
             return Expr(:try, try_clause, ex.args[2], catch_clause, finally_clause)
         elseif isexpr(ex, :let)
             letlocals = union(locals, localvars(ex.args[1]))
-            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], copy(globals), letlocals, true))
+            return Expr(ex.head, localassignment(ex.args[1], copy(globals), copy(locals), true),
```

Comment:
This doesn't seem correct to me.  I think you have to go through the `let` assignments from left to right, adding to the `letlocals` list one at a time, and using `letlocals` for `localassignment`.

In particular, consider:
```
julia> a = 17
17

julia> let a = 1, b = (a=2)+1; b; end
3

julia> a
17
```
Even though `a` is a global, by the time Julia gets to `b = a+1`, the local variable `a` has been defined, so `a=2` should not be transformed into `global a=2`.

Reply:
Yes, that is exactly what this PR does. There is even a couple of test cases for it :smile: 
```
julia> a = 17
17

julia> @softscope (let a = 1, b = (a=2)+1; b; end)
3

julia> a
17
```
The `localassignment` can deal with a single assignment or a block; in the case of a block it goes left to right as you mention. (Lines 92-99 in SoftGlobalScope.jl)

Reply:
I see, I got confused by line 131 where you separately call `localvars`